File: demos/cpp/basic_usage.c
/**
 * @file basic_usage.c
 * @brief Demonstrates basic Entangle database operations in C
 * 
 * This demo shows how to:
 * - Open and close a database
 * - Write values of all types (float, int, text, bytes)
 * - Read values back
 * - Delete keys
 * - List all keys
 * - Handle errors properly
 * 
 * Build: See Makefile in this directory
 * 
 * _Requirements: 3.1_
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/entangle.h"

/* Helper macro for error checking */
#define CHECK_ERROR(result, operation) \
    do { \
        if ((result) < ENTANGLE_SUCCESS) { \
            fprintf(stderr, "Error in %s: %d\n", operation, result); \
            return 1; \
        } \
    } while(0)

/* Helper function to print error code meaning */
static const char* error_to_string(int32_t code) {
    switch (code) {
        case ENTANGLE_SUCCESS:         return "Success";
        case ENTANGLE_ERR_GENERIC:     return "Generic error";
        case ENTANGLE_ERR_DB_NOT_FOUND: return "Database not found";
        case ENTANGLE_ERR_INVALID_PATH: return "Invalid path";
        case ENTANGLE_ERR_IO:          return "I/O error";
        case ENTANGLE_ERR_SERIALIZATION: return "Serialization error";
        case ENTANGLE_ERR_KEY_NOT_FOUND: return "Key not found";
        case ENTANGLE_ERR_TYPE_MISMATCH: return "Type mismatch";
        case ENTANGLE_ERR_EMPTY_KEY:   return "Empty key";
        case ENTANGLE_ERR_KEY_TOO_LONG: return "Key too long";
        case ENTANGLE_ERR_INTERNAL_PANIC: return "Internal panic";
        default:                       return "Unknown error";
    }
}

/* Demo: Basic CRUD operations */
static int demo_crud(const char* db_path) {
    int32_t result;
    int64_t offset;
    
    printf("\n=== Demo: Basic CRUD Operations ===\n");
    
    /* Open database */
    printf("Opening database at: %s\n", db_path);
    result = entangle_open(db_path);
    CHECK_ERROR(result, "entangle_open");
    printf("  -> Database opened successfully\n");
    
    /* Write a float value */
    printf("\nWriting float value: temperature = 23.5\n");
    offset = entangle_put_float(db_path, "temperature", 23.5);
    if (offset < 0) {
        fprintf(stderr, "  -> Error: %s\n", error_to_string((int32_t)offset));
        return 1;
    }
    printf("  -> Written at offset: %lld\n", (long long)offset);
    
    /* Write an integer value */
    printf("\nWriting int value: count = 42\n");
    offset = entangle_put_int(db_path, "count", 42);
    if (offset < 0) {
        fprintf(stderr, "  -> Error: %s\n", error_to_string((int32_t)offset));
        return 1;
    }
    printf("  -> Written at offset: %lld\n", (long long)offset);
    
    /* Write a text value */
    printf("\nWriting text value: message = \"Hello, Entangle!\"\n");
    offset = entangle_put_text(db_path, "message", "Hello, Entangle!");
    if (offset < 0) {
        fprintf(stderr, "  -> Error: %s\n", error_to_string((int32_t)offset));
        return 1;
    }
    printf("  -> Written at offset: %lld\n", (long long)offset);
    
    /* Write a bytes value */
    uint8_t binary_data[] = {0xDE, 0xAD, 0xBE, 0xEF, 0xCA, 0xFE};
    printf("\nWriting bytes value: binary_data = [0xDE, 0xAD, 0xBE, 0xEF, 0xCA, 0xFE]\n");
    offset = entangle_put_bytes(db_path, "binary_data", binary_data, sizeof(binary_data));
    if (offset < 0) {
        fprintf(stderr, "  -> Error: %s\n", error_to_string((int32_t)offset));
        return 1;
    }
    printf("  -> Written at offset: %lld\n", (long long)offset);
    
    /* Read float value back */
    double temp_value;
    printf("\nReading float value: temperature\n");
    result = entangle_get_float(db_path, "temperature", &temp_value);
    if (result == ENTANGLE_SUCCESS) {
        printf("  -> Value: %f\n", temp_value);
    } else {
        fprintf(stderr, "  -> Error: %s\n", error_to_string(result));
    }
    
    /* Read integer value back */
    int64_t count_value;
    printf("\nReading int value: count\n");
    result = entangle_get_int(db_path, "count", &count_value);
    if (result == ENTANGLE_SUCCESS) {
        printf("  -> Value: %lld\n", (long long)count_value);
    } else {
        fprintf(stderr, "  -> Error: %s\n", error_to_string(result));
    }
    
    /* Read text value back */
    size_t text_len;
    printf("\nReading text value: message\n");
    char* text_value = entangle_get_text(db_path, "message", &text_len);
    if (text_value != NULL) {
        printf("  -> Value: \"%s\" (length: %zu)\n", text_value, text_len);
        entangle_free_text(text_value, text_len);
    } else {
        fprintf(stderr, "  -> Error: Key not found or type mismatch\n");
    }
    
    /* Read bytes value back */
    size_t bytes_len;
    printf("\nReading bytes value: binary_data\n");
    uint8_t* bytes_value = entangle_get_bytes(db_path, "binary_data", &bytes_len);
    if (bytes_value != NULL) {
        printf("  -> Value: [");
        for (size_t i = 0; i < bytes_len; i++) {
            printf("0x%02X%s", bytes_value[i], i < bytes_len - 1 ? ", " : "");
        }
        printf("] (length: %zu)\n", bytes_len);
        entangle_free_bytes(bytes_value, bytes_len);
    } else {
        fprintf(stderr, "  -> Error: Key not found or type mismatch\n");
    }
    
    /* Check if key exists */
    printf("\nChecking if 'temperature' exists\n");
    result = entangle_exists(db_path, "temperature");
    printf("  -> Exists: %s\n", result == 1 ? "yes" : "no");
    
    printf("\nChecking if 'nonexistent' exists\n");
    result = entangle_exists(db_path, "nonexistent");
    printf("  -> Exists: %s\n", result == 1 ? "yes" : "no");
    
    /* List all keys */
    size_t num_keys;
    printf("\nListing all keys:\n");
    char** keys = entangle_keys(db_path, &num_keys);
    if (keys != NULL) {
        for (size_t i = 0; i < num_keys; i++) {
            printf("  -> Key %zu: %s\n", i + 1, keys[i]);
        }
        entangle_free_keys(keys, num_keys);
    } else if (num_keys == 0) {
        printf("  -> No keys found\n");
    } else {
        fprintf(stderr, "  -> Error listing keys\n");
    }
    
    /* Delete a key */
    printf("\nDeleting key: count\n");
    result = entangle_delete(db_path, "count");
    CHECK_ERROR(result, "entangle_delete");
    printf("  -> Key deleted successfully\n");
    
    /* Verify deletion */
    printf("\nVerifying deletion - reading 'count':\n");
    result = entangle_get_int(db_path, "count", &count_value);
    if (result == ENTANGLE_ERR_KEY_NOT_FOUND) {
        printf("  -> Key not found (as expected after deletion)\n");
    } else if (result == ENTANGLE_SUCCESS) {
        printf("  -> Unexpected: Key still exists with value %lld\n", (long long)count_value);
    } else {
        fprintf(stderr, "  -> Error: %s\n", error_to_string(result));
    }
    
    /* List keys after deletion */
    printf("\nListing keys after deletion:\n");
    keys = entangle_keys(db_path, &num_keys);
    if (keys != NULL) {
        for (size_t i = 0; i < num_keys; i++) {
            printf("  -> Key %zu: %s\n", i + 1, keys[i]);
        }
        entangle_free_keys(keys, num_keys);
    } else if (num_keys == 0) {
        printf("  -> No keys found\n");
    }
    
    /* Close database */
    printf("\nClosing database\n");
    result = entangle_close(db_path);
    CHECK_ERROR(result, "entangle_close");
    printf("  -> Database closed successfully\n");
    
    return 0;
}

/* Demo: Time-series data and tensor extraction */
static int demo_timeseries(const char* db_path) {
    int32_t result;
    
    printf("\n=== Demo: Time-Series and Tensor Extraction ===\n");
    
    /* Open database */
    result = entangle_open(db_path);
    CHECK_ERROR(result, "entangle_open");
    
    /* Write multiple float values to simulate time-series */
    printf("\nWriting 10 sensor readings:\n");
    for (int i = 0; i < 10; i++) {
        double value = 20.0 + (double)i * 0.5;  /* Simulated temperature */
        int64_t offset = entangle_put_float(db_path, "sensor/temp", value);
        if (offset < 0) {
            fprintf(stderr, "  -> Error writing value %d\n", i);
            return 1;
        }
        printf("  -> Reading %d: %.1f\n", i + 1, value);
    }
    
    /* Extract history as tensor */
    size_t tensor_len;
    printf("\nExtracting history tensor for 'sensor/temp':\n");
    double* tensor = entangle_get_history_tensor(db_path, "sensor/temp", &tensor_len);
    if (tensor != NULL) {
        printf("  -> Tensor length: %zu\n", tensor_len);
        printf("  -> Values: [");
        for (size_t i = 0; i < tensor_len; i++) {
            printf("%.1f%s", tensor[i], i < tensor_len - 1 ? ", " : "");
        }
        printf("]\n");
        
        /* Calculate simple statistics */
        double sum = 0.0, min = tensor[0], max = tensor[0];
        for (size_t i = 0; i < tensor_len; i++) {
            sum += tensor[i];
            if (tensor[i] < min) min = tensor[i];
            if (tensor[i] > max) max = tensor[i];
        }
        printf("  -> Mean: %.2f, Min: %.1f, Max: %.1f\n", 
               sum / tensor_len, min, max);
        
        /* Free tensor memory */
        entangle_free_tensor(tensor, tensor_len);
    } else {
        fprintf(stderr, "  -> Error extracting tensor\n");
    }
    
    /* Close database */
    result = entangle_close(db_path);
    CHECK_ERROR(result, "entangle_close");
    
    return 0;
}

/* Demo: Error handling */
static int demo_error_handling(const char* db_path) {
    int32_t result;
    
    printf("\n=== Demo: Error Handling ===\n");
    
    /* Try to read from unopened database */
    printf("\nTrying to read from unopened database:\n");
    double value;
    result = entangle_get_float(db_path, "test", &value);
    printf("  -> Result: %d (%s)\n", result, error_to_string(result));
    
    /* Open database */
    result = entangle_open(db_path);
    CHECK_ERROR(result, "entangle_open");
    
    /* Try to read non-existent key */
    printf("\nTrying to read non-existent key:\n");
    result = entangle_get_float(db_path, "nonexistent_key", &value);
    printf("  -> Result: %d (%s)\n", result, error_to_string(result));
    
    /* Write an int, try to read as float (type mismatch) */
    printf("\nWriting int, then trying to read as float:\n");
    entangle_put_int(db_path, "int_value", 100);
    result = entangle_get_float(db_path, "int_value", &value);
    printf("  -> Result: %d (%s)\n", result, error_to_string(result));
    
    /* Try with NULL pointers */
    printf("\nTrying operations with NULL pointers:\n");
    result = entangle_get_float(db_path, NULL, &value);
    printf("  -> get_float with NULL key: %d (%s)\n", result, error_to_string(result));
    
    result = entangle_get_float(db_path, "test", NULL);
    printf("  -> get_float with NULL out: %d (%s)\n", result, error_to_string(result));
    
    /* Close database */
    result = entangle_close(db_path);
    CHECK_ERROR(result, "entangle_close");
    
    return 0;
}

/* Demo: Compaction */
static int demo_compaction(const char* db_path) {
    int32_t result;
    
    printf("\n=== Demo: Database Compaction ===\n");
    
    /* Open database */
    result = entangle_open(db_path);
    CHECK_ERROR(result, "entangle_open");
    
    /* Write multiple values to same key */
    printf("\nWriting 5 values to same key 'counter':\n");
    for (int i = 1; i <= 5; i++) {
        entangle_put_int(db_path, "counter", i * 10);
        printf("  -> Write %d: %d\n", i, i * 10);
    }
    
    /* Check history before compaction */
    size_t tensor_len;
    printf("\nHistory before compaction:\n");
    double* tensor = entangle_get_history_tensor(db_path, "counter", &tensor_len);
    if (tensor != NULL) {
        printf("  -> Length: %zu values\n", tensor_len);
        entangle_free_tensor(tensor, tensor_len);
    }
    
    /* Compact database */
    printf("\nCompacting database...\n");
    result = entangle_compact(db_path);
    CHECK_ERROR(result, "entangle_compact");
    printf("  -> Compaction complete\n");
    
    /* Check history after compaction */
    printf("\nHistory after compaction:\n");
    tensor = entangle_get_history_tensor(db_path, "counter", &tensor_len);
    if (tensor != NULL) {
        printf("  -> Length: %zu value(s)\n", tensor_len);
        if (tensor_len > 0) {
            printf("  -> Latest value: %.0f\n", tensor[0]);
        }
        entangle_free_tensor(tensor, tensor_len);
    }
    
    /* Close database */
    result = entangle_close(db_path);
    CHECK_ERROR(result, "entangle_close");
    
    return 0;
}

int main(int argc, char* argv[]) {
    const char* db_path = "demo_basic.db";
    
    /* Allow custom database path */
    if (argc > 1) {
        db_path = argv[1];
    }
    
    printf("Entangle C Basic Usage Demo\n");
    printf("============================\n");
    printf("Database path: %s\n", db_path);
    
    /* Run demos */
    int result = 0;
    
    result = demo_crud(db_path);
    if (result != 0) {
        fprintf(stderr, "CRUD demo failed\n");
        return result;
    }
    
    result = demo_timeseries(db_path);
    if (result != 0) {
        fprintf(stderr, "Time-series demo failed\n");
        return result;
    }
    
    result = demo_error_handling(db_path);
    if (result != 0) {
        fprintf(stderr, "Error handling demo failed\n");
        return result;
    }
    
    result = demo_compaction(db_path);
    if (result != 0) {
        fprintf(stderr, "Compaction demo failed\n");
        return result;
    }
    
    printf("\n============================\n");
    printf("All demos completed successfully!\n");
    
    /* Cleanup: remove demo database */
    remove(db_path);
    
    return 0;
}

File: demos/python/basic_usage.py
#!/usr/bin/env python3
"""
Basic Syna Usage Demo

This demo shows fundamental database operations using the Python wrapper:
- Library loading and database open/close
- Writing values of different types (Float, Int, Text, Bytes)
- Reading values back
- Deleting keys
- Listing keys
- Error handling patterns

Requirements: 2.1 - WHEN a developer views the Python basic demo THEN the demo 
SHALL show loading the library, opening DB, and CRUD operations

Run with: python basic_usage.py
"""

import os
import sys
import tempfile

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from Syna import SynaDB, SynaError


def demo_library_loading():
    """Demonstrate library loading and path discovery."""
    print("=" * 60)
    print("1. LIBRARY LOADING")
    print("=" * 60)
    
    # The library is loaded automatically when SynaDB is instantiated
    # Show the library path that was found
    print("\nThe Syna library is loaded via ctypes from the shared library.")
    print("Supported platforms:")
    print("  - Linux:   libsynadb.so")
    print("  - macOS:   libsynadb.dylib")
    print("  - Windows: synadb.dll")
    print("\nSearch paths (in order):")
    print("  1. target/release/ (relative to package)")
    print("  2. target/debug/   (relative to package)")
    print("  3. Current working directory")
    print("  4. System library path")
    
    # Show actual library path after loading
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "test.db")
        with SynaDB(db_path) as db:
            print(f"\n✓ Library loaded from: {db._lib_path}")
    
    print()


def demo_database_open_close():
    """Demonstrate database open and close operations."""
    print("=" * 60)
    print("2. DATABASE OPEN/CLOSE")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        # Method 1: Context manager (recommended)
        print("\nMethod 1: Context Manager (recommended)")
        print("-" * 40)
        print("""
with SynaDB("my.db") as db:
    # Database is open here
    db.put_float("key", 3.14)
# Database is automatically closed here
""")
        
        with SynaDB(db_path) as db:
            db.put_float("test", 1.0)
            print("✓ Database opened with context manager")
        print("✓ Database automatically closed on exit")
        
        # Method 2: Manual open/close
        print("\nMethod 2: Manual Open/Close")
        print("-" * 40)
        print("""
db = SynaDB("my.db")
try:
    db.put_float("key", 3.14)
finally:
    db.close()
""")
        
        db = SynaDB(db_path)
        try:
            db.put_float("test2", 2.0)
            print("✓ Database opened manually")
        finally:
            db.close()
            print("✓ Database closed manually")
    
    print()


def demo_write_operations():
    """Demonstrate all write operations."""
    print("=" * 60)
    print("3. WRITE OPERATIONS (PUT)")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        with SynaDB(db_path) as db:
            # Float
            print("\nput_float(key, value) -> offset")
            print("-" * 40)
            offset = db.put_float("sensor/temperature", 23.5)
            print(f"db.put_float('sensor/temperature', 23.5)")
            print(f"✓ Written at offset: {offset}")
            
            # Integer
            print("\nput_int(key, value) -> offset")
            print("-" * 40)
            offset = db.put_int("counter/visits", 42)
            print(f"db.put_int('counter/visits', 42)")
            print(f"✓ Written at offset: {offset}")
            
            # Text
            print("\nput_text(key, value) -> offset")
            print("-" * 40)
            offset = db.put_text("config/name", "Syna Demo")
            print(f"db.put_text('config/name', 'Syna Demo')")
            print(f"✓ Written at offset: {offset}")
            
            # Bytes
            print("\nput_bytes(key, value) -> offset")
            print("-" * 40)
            data = b"\x00\x01\x02\x03\xff"
            offset = db.put_bytes("binary/data", data)
            print(f"db.put_bytes('binary/data', b'\\x00\\x01\\x02\\x03\\xff')")
            print(f"✓ Written at offset: {offset}")
            
            # Appending multiple values (history)
            print("\nAppending Multiple Values (History)")
            print("-" * 40)
            print("Each put() appends a new value, preserving history:")
            for i, temp in enumerate([24.0, 24.5, 25.0, 25.5]):
                offset = db.put_float("sensor/temperature", temp)
                print(f"  put_float('sensor/temperature', {temp}) -> offset {offset}")
    
    print()


def demo_read_operations():
    """Demonstrate all read operations."""
    print("=" * 60)
    print("4. READ OPERATIONS (GET)")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        with SynaDB(db_path) as db:
            # Setup test data
            db.put_float("sensor/temp", 23.5)
            db.put_float("sensor/temp", 24.0)
            db.put_float("sensor/temp", 24.5)
            db.put_int("counter", 100)
            db.put_text("message", "Hello, Syna!")
            db.put_bytes("binary", b"\xde\xad\xbe\xef")
            
            # get_float
            print("\nget_float(key) -> Optional[float]")
            print("-" * 40)
            value = db.get_float("sensor/temp")
            print(f"db.get_float('sensor/temp') = {value}")
            print("✓ Returns the LATEST value for the key")
            
            # get_int
            print("\nget_int(key) -> Optional[int]")
            print("-" * 40)
            value = db.get_int("counter")
            print(f"db.get_int('counter') = {value}")
            
            # get_text
            print("\nget_text(key) -> Optional[str]")
            print("-" * 40)
            value = db.get_text("message")
            print(f"db.get_text('message') = '{value}'")
            
            # get_bytes
            print("\nget_bytes(key) -> Optional[bytes]")
            print("-" * 40)
            value = db.get_bytes("binary")
            print(f"db.get_bytes('binary') = {value}")
            
            # get_history_tensor (for ML)
            print("\nget_history_tensor(key) -> np.ndarray")
            print("-" * 40)
            history = db.get_history_tensor("sensor/temp")
            print(f"db.get_history_tensor('sensor/temp') = {history}")
            print(f"✓ Shape: {history.shape}, dtype: {history.dtype}")
            print("✓ Returns ALL float values in chronological order")
            
            # Non-existent key
            print("\nReading Non-Existent Key")
            print("-" * 40)
            value = db.get_float("nonexistent")
            print(f"db.get_float('nonexistent') = {value}")
            print("✓ Returns None (not an error)")
    
    print()


def demo_delete_operations():
    """Demonstrate delete operations."""
    print("=" * 60)
    print("5. DELETE OPERATIONS")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        with SynaDB(db_path) as db:
            # Setup
            db.put_float("to_delete", 1.0)
            db.put_float("to_keep", 2.0)
            
            print("\nBefore Delete:")
            print(f"  exists('to_delete') = {db.exists('to_delete')}")
            print(f"  exists('to_keep') = {db.exists('to_keep')}")
            
            # Delete
            print("\ndelete(key)")
            print("-" * 40)
            db.delete("to_delete")
            print("db.delete('to_delete')")
            
            print("\nAfter Delete:")
            print(f"  exists('to_delete') = {db.exists('to_delete')}")
            print(f"  get_float('to_delete') = {db.get_float('to_delete')}")
            print("✓ Deleted keys return None and exists() returns False")
            
            # exists()
            print("\nexists(key) -> bool")
            print("-" * 40)
            print(f"db.exists('to_keep') = {db.exists('to_keep')}")
            print(f"db.exists('to_delete') = {db.exists('to_delete')}")
            print(f"db.exists('never_existed') = {db.exists('never_existed')}")
    
    print()


def demo_key_listing():
    """Demonstrate key listing."""
    print("=" * 60)
    print("6. KEY LISTING")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        with SynaDB(db_path) as db:
            # Setup hierarchical keys
            db.put_float("sensor/temp/room1", 22.0)
            db.put_float("sensor/temp/room2", 23.0)
            db.put_float("sensor/humidity/room1", 45.0)
            db.put_int("counter/visits", 100)
            db.put_text("config/name", "Demo")
            
            # Delete one key
            db.delete("counter/visits")
            
            print("\nkeys() -> List[str]")
            print("-" * 40)
            keys = db.keys()
            print("db.keys() =")
            for key in sorted(keys):
                print(f"  - {key}")
            print(f"\n✓ Total: {len(keys)} keys")
            print("✓ Deleted keys are NOT included")
    
    print()


def demo_error_handling():
    """Demonstrate error handling patterns."""
    print("=" * 60)
    print("7. ERROR HANDLING PATTERNS")
    print("=" * 60)
    
    print("\nSynaError Exception")
    print("-" * 40)
    print("""
SynaError has two attributes:
  - code: Integer error code
  - message: Human-readable description

Error codes:
   0: Generic error
  -1: Database not found in registry
  -2: Invalid path or UTF-8
  -3: I/O error
  -4: Serialization error
  -5: Key not found
  -6: Type mismatch
  -7: Empty key not allowed
  -8: Key too long
-100: Internal panic
""")
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        # Pattern 1: Try/except for specific errors
        print("\nPattern 1: Try/Except")
        print("-" * 40)
        print("""
try:
    with SynaDB("my.db") as db:
        db.put_float("key", value)
except SynaError as e:
    print(f"Error {e.code}: {e.message}")
""")
        
        try:
            with SynaDB(db_path) as db:
                # Try to use empty key (will fail)
                db.put_float("", 1.0)
        except SynaError as e:
            print(f"✓ Caught error: code={e.code}, message='{e.message}'")
        
        # Pattern 2: Check for None on reads
        print("\nPattern 2: Check for None on Reads")
        print("-" * 40)
        print("""
value = db.get_float("key")
if value is None:
    print("Key not found")
else:
    print(f"Value: {value}")
""")
        
        with SynaDB(db_path) as db:
            value = db.get_float("nonexistent")
            if value is None:
                print("✓ Key not found (returned None)")
            else:
                print(f"Value: {value}")
        
        # Pattern 3: Check existence before operations
        print("\nPattern 3: Check Existence First")
        print("-" * 40)
        print("""
if db.exists("key"):
    value = db.get_float("key")
    # Safe to use value
""")
        
        with SynaDB(db_path) as db:
            db.put_float("existing", 42.0)
            
            if db.exists("existing"):
                value = db.get_float("existing")
                print(f"✓ Key exists, value = {value}")
            
            if not db.exists("missing"):
                print("✓ Key does not exist, skipping read")
        
        # Pattern 4: Using closed database
        print("\nPattern 4: Handling Closed Database")
        print("-" * 40)
        
        db = SynaDB(db_path)
        db.close()
        
        try:
            db.put_float("key", 1.0)
        except SynaError as e:
            print(f"✓ Caught error on closed DB: '{e.message}'")
    
    print()


def demo_compaction():
    """Demonstrate database compaction."""
    print("=" * 60)
    print("8. DATABASE COMPACTION")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        with SynaDB(db_path) as db:
            # Write many values
            print("\nWriting 100 values to same key...")
            for i in range(100):
                db.put_float("sensor", float(i))
            
            # Check file size before
            size_before = os.path.getsize(db_path)
            print(f"File size before compaction: {size_before} bytes")
            
            # Get history
            history = db.get_history_tensor("sensor")
            print(f"History length: {len(history)} values")
            
            # Compact
            print("\ncompact()")
            print("-" * 40)
            db.compact()
            print("db.compact()")
            
            # Check file size after
            size_after = os.path.getsize(db_path)
            print(f"File size after compaction: {size_after} bytes")
            print(f"Space saved: {size_before - size_after} bytes")
            
            # History after compaction
            history = db.get_history_tensor("sensor")
            print(f"History length after compaction: {len(history)} value(s)")
            print("✓ Compaction keeps only the latest value per key")
    
    print()


def demo_persistence():
    """Demonstrate data persistence across sessions."""
    print("=" * 60)
    print("9. DATA PERSISTENCE")
    print("=" * 60)
    
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        
        # Session 1: Write data
        print("\nSession 1: Writing data")
        print("-" * 40)
        with SynaDB(db_path) as db:
            db.put_float("persistent/value", 42.0)
            db.put_text("persistent/message", "Hello from session 1!")
            print("✓ Data written and database closed")
        
        # Session 2: Read data back
        print("\nSession 2: Reading data back")
        print("-" * 40)
        with SynaDB(db_path) as db:
            value = db.get_float("persistent/value")
            message = db.get_text("persistent/message")
            print(f"persistent/value = {value}")
            print(f"persistent/message = '{message}'")
            print("✓ Data persisted across sessions!")
    
    print()


def main():
    """Run all demos."""
    print("\n" + "=" * 60)
    print("   Syna PYTHON BASIC USAGE DEMO")
    print("   Requirements: 2.1")
    print("=" * 60 + "\n")
    
    try:
        demo_library_loading()
        demo_database_open_close()
        demo_write_operations()
        demo_read_operations()
        demo_delete_operations()
        demo_key_listing()
        demo_error_handling()
        demo_compaction()
        demo_persistence()
        
        print("=" * 60)
        print("   ALL DEMOS COMPLETED SUCCESSFULLY!")
        print("=" * 60 + "\n")
        return 0
        
    except Exception as e:
        print(f"\n❌ Demo failed with error: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(main())

